Improve TextEdit and CodeEdit documentation

kitbdev opened the pull request:
- fixes https://github.com/godotengine/godot/issues/98260
- The descriptions for `get_visible_line_count` and `get_total_visible_line_count` in particular seemed to be swapped.
- Removes `Sets whether` and `Sets if`.
- Uses `[member ProjectSettings.input/...]` for actions (and found `ui_text_unindent` is actually supposed to be `ui_text_dedent`).
- Adds references to related methods, members, and theme items.
- Adds details in some places that were lacking.

kitbdev committed:
Improve TextEdit and CodeEdit documentation

Loff3 commented on doc/classes/TextEdit.xml:
```diff
@@ -1269,7 +1269,7 @@
 		</member>
 		<member name="caret_multiple" type="bool" setter="set_multiple_carets_enabled" getter="is_multiple_carets_enabled" default="true">
-			Sets if multiple carets are allowed.
+			If [code]true[/code], multiple carets can be used at once. Left clicking with [kbd]Alt[/kbd] will add a new caret. See [method add_caret] and [method get_caret_count].
```
```suggestion
			If [code]true[/code], allows multiple carets. Left-clicking with [kbd]Alt[/kbd] adds a new caret. See [method add_caret] and [method get_caret_count].
```

Loff3 commented on doc/classes/CodeEdit.xml:
```diff
@@ -500,20 +500,20 @@
 			<return type="void" />
 			<description>
-				Unindents selected lines, or in the case of no selection the caret line by one. Same as performing "ui_text_unindent" action.
+				Unindents all lines that are selected or have a caret on them. Uses spaces or a tab depending on [member indent_use_spaces]. Same as performing the [member ProjectSettings.input/ui_text_dedent] action. See [method indent_lines].
```
```suggestion
				Unindents all selected lines or lines containing the caret. Uses spaces or a tab depending on [member indent_use_spaces]. Equivalent to performing the [member ProjectSettings.input/ui_text_dedent] action. See [method indent_lines].				
```

kitbdev replied:
For `lines that are selected or have a caret on them` I made this way to match other descriptions for consistency. 
Also in your suggestion, it would be `and` instead of `or`.

Loff3 commented on doc/classes/CodeEdit.xml:
```diff
@@ -458,23 +458,23 @@
 			<param index="1" name="breakpointed" type="bool" />
 			<description>
-				Sets the line as breakpointed.
+				Sets the line as breakpointed. If [code]true[/code] and [member gutters_draw_breakpoints_gutter] is [code]true[/code], shows the [theme_item breakpoint] icon in the gutter for this line. See [method get_breakpointed_lines] and [method is_line_breakpointed].
```
```suggestion
				Sets the line as a breakpoint. If [code]true[/code] and [member gutters_draw_breakpoints_gutter] is [code]true[/code], displays the [theme_item breakpoint] icon in the gutter for this line. See [method get_breakpointed_lines] and [method is_line_breakpointed].
```

kitbdev replied:
I think I will use 'is drawn' instead of 'displays' or 'shows', to better match other mentions of icon.

Loff3 replied:
Yeah, good choice to keep the wording consistent

Loff3 reviewed (changes requested):
I've suggested a few small changes to the wording that might help improve clarity.

lostminds commented on doc/classes/CodeEdit.xml:
```diff
@@ -527,16 +527,16 @@
 	<members>
 		<member name="auto_brace_completion_enabled" type="bool" setter="set_auto_brace_completion_enabled" getter="is_auto_brace_completion_enabled" default="false">
-			Sets whether brace pairs should be autocompleted.
+			If [code]true[/code], uses [member auto_brace_completion_pairs] to automatically insert the closing brace when the opening brace is inserted by typing or autocompletion. Also automatically removes the closing brace when using backspace on the opening brace.
```
Looks good, but maybe add an explicit mention of quotes as well? As the setting also applies to closing quotes but this is not apparent from the name or description, making it hard to find if you're looking to stop the editor from inserting matching quotes. For example change to _"... insert the closing brace or quote when an opening one is inserted by... "_

kitbdev replied:
Check the Editor Settings description, this is the CodeEdit one. This will use whatever is set in the `auto_brace_completion_pairs`, so there may not be any quotations.

The Editor Settings one mentions quotations and comments.

https://github.com/godotengine/godot/pull/97656/files#diff-52f0c165415d19abd2d5586c8a168169a8d39716be86688a9117125409b421f0

lostminds replied:
Right, I wasn't aware of the auto_brace_completion_pairs setting. Then I agree that your original wording is a better choice, even if this design choice with a configurable auto_brace_completion_pairs set and the naming of the settings makes it hard to find and understand these settings if you're looking to control the behavior regarding quotes. But this might then be an issue beyond this PR, and it's still a big improvement over the previous state, thanks!

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -153,7 +153,7 @@
 			<return type="void" />
 			<description>
-				Perform an indent as if the user activated the "ui_text_indent" action.
+				Perform an indent as if the user activated the [member ProjectSettings.input/ui_text_indent] action. If there is a selection this indents lines like [method indent_lines], otherwise indentation is added at the caret. Uses spaces or a tab depending on [member indent_use_spaces].
```
```suggestion
				Perform an indent as if the user activated the [member ProjectSettings.input/ui_text_indent] action. If there is a selection, the selected lines are indented like [method indent_lines], otherwise indentation is added at the caret. Uses multiple spaces or a tab character depending on [member indent_use_spaces].
```
Or:
```suggestion
				Indents as if the user activated the [member ProjectSettings.input/ui_text_indent] action. If lines are selected, they are indented like [method indent_lines]. Otherwise, The additional indentation is added at the caret. The used characters depend on [member indent_use_spaces].
```
Or similar. Feel free to contest.

Mickeon reviewed (commented):
Thank you very, very much for changing these for the better. These descriptions were written really awkwardly, but there could be work to do still.

kitbdev commented on doc/classes/CodeEdit.xml:
```diff
@@ -579,7 +579,8 @@
 		<member name="layout_direction" type="int" setter="set_layout_direction" getter="get_layout_direction" overrides="Control" enum="Control.LayoutDirection" default="2" />
 		<member name="line_folding" type="bool" setter="set_line_folding_enabled" getter="is_line_folding_enabled" default="false">
-			Sets whether line folding is allowed.
+			If [code]false[/code], line folding methods like [method fold_line] will not work and [method can_fold_line] will always return [code]false[/code].
+			If [code]true[/code], lines can be folded. See [member gutters_draw_fold_gutter].
```
I have it this way since false is default, but I can switch it if wanted.

Mickeon replied:
It's not a deal breaker. The current description **is** pretty good and the splitting looks nice. Perhaps it could be like this, instead:
```suggestion
			If [code]true[/code], lines can be folded. Otherwise, line folding methods like [method fold_line] will not work and [method can_fold_line] will always return [code]false[/code]. See [member gutters_draw_fold_gutter].
```

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -656,7 +656,7 @@
 			<return type="int" />
 			<description>
-				Returns the number of lines that may be drawn.
+				Returns the number of lines in the text, including wrapped lines and excluding folded lines. If [member wrap_mode] is set to [constant LINE_WRAPPING_NONE] and no lines are folded (see [method CodeEdit.is_line_folded]) then this is equivalent to [method get_line_count]. See [method get_visible_line_count_in_range] for a limited range of lines.
```
I want to avoid `may be drawn` since the lines it counts may not be visible (ex. lines that to be scrolled down to).
Also it sort of implies to be the number of lines that fit in the control, but there can be less than that.

Mickeon replied:
I see. My main issue is that there's no summary of what this method considers. The function signature is not localized, so it can only be relied upon so much.
But it's not that bad, saying "total" may be just fine:
```suggestion
				Returns the total number of lines in the text. This includes wrapped lines and excludes folded lines. See [method get_visible_line_count_in_range] for a limited range of lines.
				[b]Note:[/b] If [member wrap_mode] is set to [constant LINE_WRAPPING_NONE] and no lines are folded (see [method CodeEdit.is_line_folded]) then this is equivalent to [method get_line_count].
```

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -674,15 +674,15 @@
 			<param index="1" name="to_line" type="int" />
 			<description>
-				Returns the total number of visible + wrapped lines between the two lines.
+				Returns the total number of visible lines (including wrapped text) between the [param from_line] and [param to_line], inclusive. If the range covers all lines it is equivalent to [method get_total_visible_line_count].
```
I wanted to specify `get_total_visible_line_count`, to help prevent confusion with `get_visible_line_count`.

Also adjusted to `Returns the total number of lines, including wrapped lines and excluding folded lines, between [param from_line] and [param to_line] (inclusive).`, to be more accurate.

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -1045,7 +1045,7 @@
 			<param index="1" name="overwritable" type="bool" />
 			<description>
-				Sets the gutter to overwritable. See [method merge_gutters].
+				If [code]true[/code], the gutter lines can be overridden when using [method merge_gutters]. See [method is_gutter_overwritable].
```
"overwritable" means "can be overridden", so I'm not sure this suggestion is right. Also I am trying to describe the scenario when may be overridden.  When true it is always "overwitable".

I'll change "the gutter lines" to "lines of the gutter at the given index".

kitbdev commented:
Rebased and applied suggestions.
I think I got everywhere that I already touched for the gutter at index or given line. I changed a few others, but there is more that I didn't change.
Also added descriptions for a few more methods like `CodeEdit::can_fold_line` and `TextEdit::get_indent_level`.

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -527,16 +527,16 @@
 		</member>
 		<member name="auto_brace_completion_highlight_matching" type="bool" setter="set_highlight_matching_braces_enabled" getter="is_highlight_matching_braces_enabled" default="false">
-			Highlight mismatching brace pairs.
+			If [code]true[/code], highlights brace pairs when the caret is on either one using [member auto_brace_completion_pairs]. If matching, the pairs will be underlined. If a brace is unmatched, it is colored with [theme_item brace_mismatch_color].
```
```suggestion
			If [code]true[/code], highlights brace pairs when the caret is on either one, using [member auto_brace_completion_pairs]. If matching, the pairs will be underlined. If a brace is unmatched, it is colored with [theme_item brace_mismatch_color].
```

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -527,16 +527,16 @@
 		</member>
 		<member name="auto_brace_completion_pairs" type="Dictionary" setter="set_auto_brace_completion_pairs" getter="get_auto_brace_completion_pairs" default="{ &quot;\&quot;&quot;: &quot;\&quot;&quot;, &quot;&apos;&quot;: &quot;&apos;&quot;, &quot;(&quot;: &quot;)&quot;, &quot;[&quot;: &quot;]&quot;, &quot;{&quot;: &quot;}&quot; }">
-			Sets the brace pairs to be autocompleted.
+			Sets the brace pairs to be autocompleted. The key is the the opening brace and the value is the closing brace that matches it. A brace is a [String] made of symbols. See [member auto_brace_completion_enabled] and [member auto_brace_completion_highlight_matching].
```
```suggestion
			Sets the brace pairs to be autocompleted. For each entry in the dictionary, the key is the the opening brace and the value is the closing brace that matches it. A brace is a [String] made of symbols. See [member auto_brace_completion_enabled] and [member auto_brace_completion_highlight_matching].
```

kitbdev replied:
Also double "the" in "key is the the opening brace".

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -527,16 +527,16 @@
 		</member>
 		<member name="code_completion_enabled" type="bool" setter="set_code_completion_enabled" getter="is_code_completion_enabled" default="false">
-			Sets whether code completion is allowed.
+			If [code]true[/code], requests code completion when the [member ProjectSettings.input/ui_text_completion_query] input action is pressed. See [method _request_code_completion] or [signal code_completion_requested] to handle it.
```
Again, just a matter of consistency.
```suggestion
			If [code]true[/code], requests code completion when the [member ProjectSettings.input/ui_text_completion_query] input action is activated. To handle it, see [method _request_code_completion] or [signal code_completion_requested].
```
or:
```suggestion
			If [code]true[/code], code completion can be requested when activating the [member ProjectSettings.input/ui_text_completion_query] action. To handle it, see [method _request_code_completion] or [signal code_completion_requested].
```
Or similar?

kitbdev replied:
How about

```suggestion
			If [code]true[/code], the [member ProjectSettings.input/ui_text_completion_query] action requests code completion. To handle it, see [method _request_code_completion] or [signal code_completion_requested].
```

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -548,28 +548,28 @@
 		</member>
 		<member name="gutters_zero_pad_line_numbers" type="bool" setter="set_line_numbers_zero_padded" getter="is_line_numbers_zero_padded" default="false">
-			Sets if line numbers drawn in the gutter are zero padded.
+			If [code]true[/code], line numbers drawn in the gutter are zero padded based on the highest number. [member gutters_draw_line_numbers] must be [code]true[/code].
```
Maybe...? Up to your discretion.
```suggestion
			If [code]true[/code], line numbers drawn in the gutter are zero padded based on the total line count. [member gutters_draw_line_numbers] must be [code]true[/code].
```
Not a fan of "must be [code]true[/code]" because it doesn't really explain it, although sure, you can infer it.

kitbdev replied:
I could have the last sentence be "Line number gutters must be enabled, see [member gutters_draw_line_numbers]." instead.

Mickeon replied:
Actually on a second look the following is common enough in the class reference:
```suggestion
			If [code]true[/code], line numbers drawn in the gutter are zero padded based on the total line count. Requires [member gutters_draw_line_numbers] to be [code]true[/code].
```
or 
```suggestion
			If [code]true[/code], line numbers drawn in the gutter are zero padded based on the total line count. Requires [member gutters_draw_line_numbers] to be set to [code]true[/code].
```

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -548,28 +548,28 @@
 		</member>
 		<member name="indent_automatic" type="bool" setter="set_auto_indent_enabled" getter="is_auto_indent_enabled" default="false">
-			Sets whether automatic indent are enabled, this will add an extra indent if a prefix or brace is found.
+			If [code]true[/code], an extra indent will automatically be added when adding a new line if a prefix in [member indent_automatic_prefixes] is found. If a brace pair opening key is found (see [member auto_brace_completion_pairs]), the matching closing brace will be moved to another new line.
```
I suggest moving this parenthesis for more comfortable reading.
```suggestion
			If [code]true[/code], an extra indent will automatically be added when a new line is added and a prefix in [member indent_automatic_prefixes] is found. If a brace pair opening key is found, the matching closing brace will be moved to another new line (see [member auto_brace_completion_pairs]).
```
The verb form change is not that big of a deal. It could even be wrong or have make it worse.

kitbdev replied:
Adjusted the first part to `an extra indent is automatically inserted when a new line is added and a prefix...`.

Mickeon commented on doc/classes/CodeEdit.xml:
```diff
@@ -548,28 +548,28 @@
 		</member>
 		<member name="indent_automatic_prefixes" type="String[]" setter="set_auto_indent_prefixes" getter="get_auto_indent_prefixes" default="[&quot;:&quot;, &quot;{&quot;, &quot;[&quot;, &quot;(&quot;]">
-			Prefixes to trigger an automatic indent.
+			Prefixes to trigger an automatic indent. [member indent_automatic] must be [code]true[/code].
```
Not a fan of "must be [code]true[/code]" again, but I have no suggestions.

kitbdev replied:
For this it could just be "Used for [member indent_automatic]".

Mickeon replied:
"Used when [member indent_automatic] is set to [code]true[/code]" is my final offer

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -1275,7 +1275,7 @@
 		</member>
 		<member name="text_editor/completion/auto_brace_complete" type="bool" setter="" getter="">
-			If [code]true[/code], automatically completes braces when making use of code completion.
+			If [code]true[/code], automatically inserts the matching closing brace when the opening brace is inserted by typing or autocompletion. Also automatically removes the closing brace when using backspace on the opening brace. This includes brackets [code]() {} [][/code], string quotation marks [code]'' ""[/code], and comments [code]/**/[/code] if the language supports it.
```
```suggestion
			If [code]true[/code], automatically inserts the matching closing brace when the opening brace is inserted by typing or autocompletion. Also automatically removes the closing brace when pressing backspace on the opening brace. This includes brackets ([code]()[/code], [code][][/code], [code]{}[/code]), string quotation marks ([code]''[/code], [code]""[/code]), and comments ([code]/**/[/code]) if the language supports it.
```

Mickeon replied:
That too. `[kbd]Backspace[/kbd]`

Mickeon reviewed (commented):
Get ready for a huge second pass revision by yours truly just to ensure we don't have to do this again for a loooong while.

I am really happy about this and I must congratulate you for the hard work.

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -673,15 +673,15 @@
 			<return type="int" />
 			<description>
-				Returns the number of visible lines, including wrapped text.
+				Returns the maximum number of lines that can visually fit in the control, rounded down.
```
Maybe somethings like "Returns the number of lines that can visually fit, rounded down, based on this control's height."
Since it only really uses the height, not the whole rect.

Mickeon replied:
That's good

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -343,7 +343,7 @@
 			<param index="0" name="line" type="int" />
 			<description>
-				Returns the current background color of the line. [code]Color(0, 0, 0, 0)[/code] is returned if no color is set.
+				Returns the current background color of the given line. [code]Color(0, 0, 0, 0)[/code] is returned if no color is set.
```
"additional" makes it sound like there is multiple background colors for each line. Maybe we can say "custom background color" and mention on the setter that is is applied on top of the default background color if it is transparent?

Mickeon replied:
Sounds good to me.

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -1111,7 +1111,7 @@
 			<param index="2" name="clickable" type="bool" />
 			<description>
-				If [param clickable] is [code]true[/code], makes the [param gutter] on [param line] clickable. See [signal gutter_clicked].
+				If [param clickable] is [code]true[/code], makes the [param gutter] on [param line] clickable. This is like [method set_gutter_clickable], but for a single line. If [method is_gutter_clickable] is [code]true[/code], this will not have an effect. See [method is_line_gutter_clickable] and [signal gutter_clicked].
```
I'm a little confused on when to use the param tag. Should we make it consistent so every `given line` becomes `given [param line]`, or is it fine as is?

Mickeon replied:
I'd personally use `[param]` tags every time, but sometimes they make localization more laborious as they cannot be translated. I think in this case it looks fine.
There's unfortunately no general rules or consistency about it, and some descriptions miss `[param]` just because they were written before the tag was introduced.

I have this in my own notes:
> - Avoid using the [param] tag as a replacement for an important subject of the sentence. Ideally, rephrase the sentence as if the parameter names were generic, single-letter words. This makes descriptions easier to localize, while still keeping the [param] tags. Note that parameter names remain unlocalized.

Mickeon commented on doc/classes/TextEdit.xml:
```diff
@@ -311,7 +311,7 @@
 			<param index="0" name="line" type="int" />
 			<description>
-				Returns the number of spaces and [code]tab * tab_size[/code] before the first char.
+				Returns the indent level of the given line. This is the number of spaces that fit before the first character that is not a space or a tab character. This includes the number of actual spaces and the number of tabs multiplied by the tab size (see [method get_tab_size]).
```
I'm not particularly confident on this, myself:
```suggestion
				Returns the indent level of the given line. This is the number of spaces and tabs at the beginning of the line, with the number of tabs multiplied by the tab size (see [method get_tab_size]).
```

kitbdev replied:
Yeah, I wanted to somehow define "indent level" but it was getting awkward.
Maybe

```suggestion
				Returns the indent level of the given line. This is the number of spaces and tabs at the beginning of the line, with the tabs taking the tab size into account (see [method get_tab_size]).
```

Mickeon replied:
Sounds like an improvement to me

kitbdev commented on doc/classes/TextEdit.xml:
```diff
@@ -1154,7 +1154,7 @@
 			<param index="0" name="enabled" type="bool" />
 			<description>
-				If [code]true[/code], sets the user into overtype mode. When the user types in this mode, it will override existing text.
+				If [code]true[/code], sets the user into overtype mode. When the user types in this mode, it will override existing text. The [member ProjectSettings.input/ui_text_toggle_insert_mode] action toggles overtype mode. See [method is_overtype_mode_enabled].
```
I think "it" is referring to the TextEdit, not the user. Since the user types, but the TextEdit is what overrides existing text.
That sentence can be rewritten to something like:
"In this mode, typing overrides existing text instead of inserting text."

Mickeon replied:
The getter equivalent of this method also refers to "the user" so if you go that route that should be changed too.

kitbdev commented on doc/classes/CodeEdit.xml:
```diff
@@ -500,20 +500,20 @@
 			<return type="void" />
 			<description>
-				Unindents selected lines, or in the case of no selection the caret line by one. Same as performing "ui_text_unindent" action.
+				Unindents all lines that are selected or have a caret on them. Uses spaces or a tab depending on [member indent_use_spaces]. Equivalent to performing the [member ProjectSettings.input/ui_text_dedent] action. See [method indent_lines].
```
"activating" sound a little repetitive with the "action" after it. Also its a shortcut, so "performing a shortcut" feels better than "activating a shortcut" to me. 
Actually, "using a shortcut" is probably better, I can change both to be more consistent with "using".

Mickeon replied:
Not particularly fond of it, personally. Someone else may beg to differ. "using" is way too generic.

kitbdev replied:
How about "Equivalent to the [member ProjectSettings.input/ui_text_dedent] action"? I don't think a verb there is actually needed.
And I can reword `do_indent` to be consistent.

Mickeon replied:
Good point. In this case it's like saying "[This method is] equivalent to the [member ProjectSettings.input/ui_text_dedent] action". Although one could argue it's ambiguous since one is a method and one is a property, but... n-eh...

Mickeon reviewed (approved):
No beating around the bush. All of this is a **huge** improvement.
